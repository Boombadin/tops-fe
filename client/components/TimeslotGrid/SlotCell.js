import cx from 'classnames';
import { isEmpty, noop } from 'lodash';
import pt from 'prop-types';
import React from 'react';

import {
  ELEMENT_ACTION,
  ELEMENT_TYPE,
  generateTestId,
} from '@client/utils/generateElementId';
export const TIME_UNAVAILABLE = 'Full';

function getClassName(props) {
  // console.log('enabled1', !props.slot.available);

  return cx(props.intervalCellClassName || 'timeslot-interval-cell', {
    disabled:
      !props.slot.enabled || !props.slot.available || !props.slot.isAllow,
    active: props.active,
    drawInterval: props.drawInterval,
  });
}

function getOnClickHandler(props) {
  return !props.slot.enabled || !props.slot.available || !props.slot.isAllow
    ? noop
    : props.onClick;
}

function getCostValue(props) {
  if (!props.slot.enabled || !props.slot.isAllow) {
    return '-';
  } else if (!props.slot.available) {
    return props.translate('timeslot.grid.slot.full');
  } else if (props.drawInterval) {
    return props.intervalName;
  } else if (props.slot.cost === 0) {
    return props.translate('timeslot.grid.slot.free');
  }
  return `${props.slot.cost} ${props.translate('baht_sign')}`;
}

/* eslint-disable jsx-a11y/click-events-have-key-events,
jsx-a11y/no-noninteractive-element-interactions,
react/no-unused-prop-types */
const SlotCell = props => (
  <div
    data-testid={generateTestId({
      type: ELEMENT_TYPE.INFO,
      action: ELEMENT_ACTION.VIEW,
      moduleName: 'SlotCell',
      uniqueId: 'Container',
    })}
    className={getClassName(props)}
    role="cell"
    onClick={getOnClickHandler(props)}
  >
    {props.active && (
      <div
        className={
          props.intervalCellActiveClassName || 'timeslot-interval-cell__active'
        }
      >
        {isEmpty(props.intervalCellCheckMarkClassName) ? (
          <div className={'timeslot-interval-cell__check-mark'}>âœ”</div>
        ) : (
          <div className={props.intervalCellCheckMarkClassName}>
            {props.translate('timeslot.grid.slot.selected')}
          </div>
        )}
      </div>
    )}
    {getCostValue(props)}
  </div>
);

SlotCell.propTypes = {
  intervalName: pt.oneOfType([pt.string, pt.object]),
  drawInterval: pt.bool.isRequired,
  active: pt.bool,
  slot: pt.shape({
    id: pt.oneOfType([pt.number, pt.string]),
    cost: pt.number,
    available: pt.bool,
    enabled: pt.bool,
  }).isRequired,
  onClick: pt.func.isRequired,
  translate: pt.func.isRequired,
  intervalCellClassName: pt.string,
  intervalCellActiveClassName: pt.string,
  intervalCellCheckMarkClassName: pt.string,
};

SlotCell.defaultProps = {
  active: false,
  intervalName: '',
  intervalCellClassName: '',
  intervalCellActiveClassName: '',
  intervalCellCheckMarkClassName: '',
};

export default SlotCell;
