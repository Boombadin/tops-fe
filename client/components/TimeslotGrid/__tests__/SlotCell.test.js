import { render, screen } from '@testing-library/react';
import React from 'react';

import SlotCell from '../SlotCell';
describe('components/TimeslotGrid/SlotCell', () => {
  const props = {
    translate: jest.fn(props => props),
    slot: {
      id: '2020-10-31/203',
      timeFrom: '10:00',
      timeTo: '12:00',
      cost: 60,
      quota: 200,
      available: true,
      enabled: true,
      isAllow: true,
    },
    drawInterval: false,
    intervalName: '',
    active: true,
    onClick: jest.fn(props => props),
  };
  test('it should match with snapshot', () => {
    const { getByTestId } = render(<SlotCell {...props} />);
    expect(getByTestId('inf-viewSlotCell-Container')).toMatchSnapshot();
    expect(screen.getByText('60 baht_sign')).toBeInTheDocument();
  });
  test('active slot | it should show text (Selected) ', () => {
    const newProps = {
      ...props,
      intervalCellCheckMarkClassName: 'custom-timeslot-interval-cell',
    };
    render(<SlotCell {...newProps} />);
    expect(screen.getByText('timeslot.grid.slot.selected')).toBeInTheDocument();
  });
  test('slot not enabled or not allow | it should show text (-) ', () => {
    const newProps = {
      ...props,
      slot: {
        id: '2020-10-31/203',
        timeFrom: '10:00',
        timeTo: '12:00',
        cost: 60,
        quota: 200,
        available: true,
        enabled: false,
        isAllow: false,
      },
    };
    render(<SlotCell {...newProps} />);
    expect(screen.getByText('-')).toBeInTheDocument();
  });
  test('slot not available | it should show text (Full) ', () => {
    const newProps = {
      ...props,
      slot: {
        id: '2020-10-31/203',
        timeFrom: '10:00',
        timeTo: '12:00',
        cost: 60,
        quota: 200,
        available: false,
        enabled: true,
        isAllow: true,
      },
    };
    render(<SlotCell {...newProps} />);
    expect(screen.getByText('timeslot.grid.slot.full')).toBeInTheDocument();
  });
  test('cost of slot = 0 | it should show text (Free) ', () => {
    const newProps = {
      ...props,
      slot: {
        id: '2020-10-31/203',
        timeFrom: '10:00',
        timeTo: '12:00',
        cost: 0,
        quota: 200,
        available: true,
        enabled: true,
        isAllow: true,
      },
    };
    render(<SlotCell {...newProps} />);
    expect(screen.getByText('timeslot.grid.slot.free')).toBeInTheDocument();
  });
  test('drawInterval is true | it should show text from intervalName (10:00-12:00) ', () => {
    const newProps = {
      ...props,
      drawInterval: true,
      intervalName: '10:00-12:00',
      slot: {
        id: '2020-10-31/203',
        timeFrom: '10:00',
        timeTo: '12:00',
        cost: 60,
        quota: 200,
        available: true,
        enabled: true,
        isAllow: true,
      },
    };
    render(<SlotCell {...newProps} />);
    expect(screen.getByText('10:00-12:00')).toBeInTheDocument();
  });
});
