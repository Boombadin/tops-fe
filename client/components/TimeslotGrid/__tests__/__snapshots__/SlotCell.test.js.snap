// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TimeslotGrid/SlotCell it should match with snapshot 1`] = `
<div
  class="timeslot-interval-cell active"
  data-testid="inf-viewSlotCell-Container"
  role="cell"
>
  <div
    class="timeslot-interval-cell__active"
  >
    <div
      class="timeslot-interval-cell__check-mark"
    >
      âœ”
    </div>
  </div>
  60 baht_sign
</div>
`;
